### PR TITLE
Use the new code generator API for Cranelift backend.

### PR DESCRIPTION
**Pull request by bjfish**
Refactor for Cranelift backend to use new code generator API
- [x] Add initial stubbed code generators
- [x] Implement code generators
- [x] Remove old compiler code
- [x] Re-enable warnings
- [x] Cleanup expect/unwrap debug usages

### CODE REVIEW

**Comment by bjfish** [comment]
bors try

**Comment by bjfish** [comment]
bors try

File: lib/runtime-abi/Cargo.toml [hunk 1]
```diff
@@ -13,7 +13,7 @@ wasmer-runtime-core = { path = "../runtime-core" }
 hashbrown = "0.1"
 failure = "0.1"
 tar = "0.4"
-wasmparser = "0.30.0"
+wasmparser = "0.29.2"
```
Comment (syrusakbary) [inline lib/runtime-abi/Cargo.toml]: Is this intended?

File: lib/runtime-core/Cargo.toml [hunk 2]
```diff
@@ -10,7 +10,7 @@ edition = "2018"
 [dependencies]
 nix = "0.12.0"
 page_size = "0.4.1"
-wasmparser = "0.30.0"
+wasmparser = "0.29.2"
```
Comment (syrusakbary) [inline lib/runtime-core/Cargo.toml]: Same. Shouldn't we use the latest version?

File: lib/runtime-core/src/parse.rs [hunk 3]
```diff
@@ -183,45 +196,54 @@ pub fn read_module<
...
             }
             ParserState::BeginFunctionBody { .. } => {
                 let id = func_count.wrapping_add(1);
                 func_count = id;
                 if func_count == 0 {
-                    info.namespace_table = namespace_builder.take().unwrap().finish();
-                    info.name_table = name_builder.take().unwrap().finish();
-                    mcg.feed_signatures(info.signatures.clone())
+                    info.write().unwrap().namespace_table =
+                        namespace_builder.take().expect("214").finish();
```
Comment (syrusakbary) [inline lib/runtime-core/src/parse.rs]: Is this expect a debug?

File: lib/runtime-core/src/parse.rs [hunk 4]
```diff
@@ -183,45 +196,54 @@ pub fn read_module<
...
                         *info
+                            .read()
+                            .unwrap()
                             .func_assoc
-                            .get(FuncIndex::new(id as usize + info.imported_functions.len()))
-                            .unwrap(),
+                            .get(FuncIndex::new(
+                                id as usize + info.read().unwrap().imported_functions.len(),
+                            ))
+                            .expect("242"),
```
Comment (syrusakbary) [inline lib/runtime-core/src/parse.rs]: Should we change this number with something easier to debug?

File: lib/runtime-abi/Cargo.toml (hunk 1 above)
Comment (bjfish) [reply lib/runtime-abi/Cargo.toml]: Yes, it's to match the current version cranelift is using, compiler will complain when different versions are used together.

File: lib/runtime-core/Cargo.toml (hunk 2 above)
Comment (bjfish) [reply lib/runtime-core/Cargo.toml]: See: https://github.com/wasmerio/wasmer/pull/437#discussion_r287612425
Trying to minimize differences with cranelift/cranelift master, we should update when they do.

File: lib/runtime-core/src/parse.rs (hunk 3 above)
Comment (bjfish) [reply lib/runtime-core/src/parse.rs]: I'll add these to the todo list for cleanup.

File: lib/runtime-core/src/parse.rs (hunk 4 above)
Comment (bjfish) [reply lib/runtime-core/src/parse.rs]: I'll change it back to unwrap as it was before, or I'll add an expect here in case we hit some potential bug.

**Comment by bjfish** [comment]
bors try

**Comment by bjfish** [comment]
bors try

File: lib/clif-backend/Cargo.toml [hunk 5]
```diff
@@ -9,13 +9,14 @@ edition = "2018"
...
 wasmer-runtime-core = { path = "../runtime-core", version = "0.4.2" }
-cranelift-native = "0.30.0"
-cranelift-codegen = "0.30.0"
-cranelift-entity = "0.30.0"
-cranelift-wasm = "0.30.0"
+cranelift-native = { git = "https://github.com/wasmerio/cranelift.git", branch = "wasmer" }
+cranelift-codegen = { git = "https://github.com/wasmerio/cranelift.git", branch = "wasmer" }
+cranelift-entity = { git = "https://github.com/wasmerio/cranelift.git", branch = "wasmer" }
+cranelift-frontend = { git = "https://github.com/wasmerio/cranelift.git", branch = "wasmer" }
+cranelift-wasm = { git = "https://github.com/wasmerio/cranelift.git", branch = "wasmer" }
```
Comment (syrusakbary) [inline lib/clif-backend/Cargo.toml]: We will probably need to update https://github.com/wasmerio/cranelift to the latest commits in cranelift

**Comment by bjfish** [comment]
bors try

**Comment by bjfish** [comment]
bors try

**Comment by bjfish** [comment]
bors try

**Comment by bjfish** [comment]
bors try

**Comment by bjfish** [comment]
bors try

**Comment by bjfish** [comment]
bors try

**Pull request merged**